File: DataStructutreAndAlgorithm/DataStructure/01_Array_ArrayList.c
/**
 * DATA STRUCTURE: Array / Dynamic Array
 * 
 * STORY:
 * Array là cơ bản nhất. Khi bạn cần lưu nhiều dữ liệu cùng kiểu,
 * thay vì tạo 100 biến riêng, bạn tạo 1 mảng 100 phần tử.
 * 
 * Tại sao quan trọng?
 * - Trong network: lưu packet buffers, socket tables
 * - Truy cập: O(1) - cực nhanh
 * - Cạn: không thể thay đổi kích thước sau khi tạo
 * 
 * Giải pháp: Dynamic Array (ArrayList)
 * Khi hết chỗ, cấp phát bộ nhớ lớn hơn và copy dữ liệu cũ vào.
 * 
 * KIẾN TRÚC:
 * ┌─────────────────────────┐
 * │ data (con trỏ)          │  → liên tục trong RAM
 * │ capacity (kích thước)   │  → bao nhiêu phần tử có thể lưu
 * │ size (số phần tử hiện)  │  → bao nhiêu phần tử đang dùng
 * └─────────────────────────┘
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// ============================================================
// 1. ĐỊNH NGHĨA CẤU TRÚC
// ============================================================

typedef struct {
    int *data;           // Con trỏ tới mảng nguyên số
    int size;            // Số phần tử hiện có
    int capacity;        // Sức chứa hiện tại
} DynamicArray;

// ============================================================
// 2. KHỞI TẠO
// ============================================================

/**
 * Tạo một dynamic array trống
 * 
 * Bước 1: Cấp phát bộ nhớ cho cấu trúc (struct)
 * Bước 2: Cấp phát bộ nhớ cho mảng dữ liệu ban đầu (capacity = 10)
 * Bước 3: Khởi tạo size = 0
 */
DynamicArray* createArray() {
    DynamicArray *arr = (DynamicArray *)malloc(sizeof(DynamicArray));
    
    arr->capacity = 10;
    arr->size = 0;
    arr->data = (int *)malloc(arr->capacity * sizeof(int));
    
    printf("✓ Tạo Array xong. Capacity: %d, Size: %d\n", arr->capacity, arr->size);
    return arr;
}

// ============================================================
// 3. RESIZE - ĐÁY LÀ CÁI CÓ PHẦN KHÓ
// ============================================================

/**
 * Khi hết chỗ, phải resize
 * 
 * Mô phỏng:
 * Trước: [1, 2, 3, _, _, _, _, _, _, _]  (capacity = 10, size = 3)
 * 
 * Sau:   [1, 2, 3, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _]
 *        (capacity = 20, size = 3)
 * 
 * Chi phí: O(n) vì phải copy tất cả phần tử cũ
 * Nhưng để amortized O(1), ta tăng gấp đôi, không tăng từng cái một
 */
void resize(DynamicArray *arr) {
    printf("  → Resize! Capacity cũ: %d\n", arr->capacity);
    
    // Tăng capacity gấp đôi
    int newCapacity = arr->capacity * 2;
    
    // Cấp phát bộ nhớ mới
    int *newData = (int *)malloc(newCapacity * sizeof(int));
    
    // Copy dữ liệu cũ sang bộ nhớ mới
    memcpy(newData, arr->data, arr->size * sizeof(int));
    
    // Giải phóng bộ nhớ cũ
    free(arr->data);
    
    // Cập nhật struct
    arr->data = newData;
    arr->capacity = newCapacity;
    
    printf("  → Resize xong. Capacity mới: %d\n", arr->capacity);
}

// ============================================================
// 4. THÊM PHẦN TỬ VÀO CUỐI
// ============================================================

/**
 * Push: Thêm phần tử vào cuối
 * 
 * if size == capacity:
 *     resize()
 * data[size] = value
 * size++
 * 
 * Độ phức tạp:
 * - Trường hợp bình thường: O(1)
 * - Khi phải resize: O(n)
 * - Amortized (trung bình): O(1)
 */
void push(DynamicArray *arr, int value) {
    // Nếu hết chỗ, phải resize
    if (arr->size == arr->capacity) {
        resize(arr);
    }
    
    // Thêm phần tử vào cuối
    arr->data[arr->size] = value;
    arr->size++;
    
    printf("  + Thêm %d. Size hiện: %d\n", value, arr->size);
}

// ============================================================
// 5. TRUY CẬP
// ============================================================

/**
 * Get: Lấy phần tử tại vị trí i
 * 
 * Độ phức tạp: O(1)
 * Vì RAM là tuần tự, biết địa chỉ đầu + offset là tìm được ngay
 */
int get(DynamicArray *arr, int index) {
    if (index < 0 || index >= arr->size) {
        printf("⚠ Lỗi: Index %d ngoài phạm vi [0, %d)\n", index, arr->size);
        return -1;
    }
    return arr->data[index];
}

// ============================================================
// 6. XOÁ PHẦN TỬ
// ============================================================

/**
 * Pop: Xóa phần tử cuối cùng
 * 
 * size--
 * 
 * Độ phức tạp: O(1)
 */
int pop(DynamicArray *arr) {
    if (arr->size == 0) {
        printf("⚠ Lỗi: Array rỗng\n");
        return -1;
    }
    
    int value = arr->data[arr->size - 1];
    arr->size--;
    
    printf("  - Xóa %d. Size hiện: %d\n", value, arr->size);
    return value;
}

// ============================================================
// 7. XOÁ TẠI VỊ TRÍ
// ============================================================

/**
 * Remove: Xóa phần tử tại vị trí index
 * 
 * Mô phỏng (xóa index=2):
 * [1, 2, 3, 4, 5]  →  [1, 2, 4, 5]
 * 
 * Phải dịch tất cả phần tử sau nó về phía trước
 * Chi phí: O(n) trong trường hợp xấu nhất (xóa ở đầu)
 */
void removeAt(DynamicArray *arr, int index) {
    if (index < 0 || index >= arr->size) {
        printf("⚠ Lỗi: Index %d ngoài phạm vi\n", index);
        return;
    }
    
    // Dịch tất cả phần tử sau index về trước 1 vị trí
    for (int i = index; i < arr->size - 1; i++) {
        arr->data[i] = arr->data[i + 1];
    }
    
    arr->size--;
    printf("  - Xóa tại index %d. Size hiện: %d\n", index, arr->size);
}

// ============================================================
// 8. IN RA
// ============================================================

void printArray(DynamicArray *arr) {
    printf("Array: [");
    for (int i = 0; i < arr->size; i++) {
        printf("%d", arr->data[i]);
        if (i < arr->size - 1) printf(", ");
    }
    printf("] (Size: %d, Capacity: %d)\n", arr->size, arr->capacity);
}

// ============================================================
// 9. GIẢI PHÓNG BỘ NHỚ
// ============================================================

void freeArray(DynamicArray *arr) {
    free(arr->data);
    free(arr);
    printf("✓ Giải phóng Array xong\n");
}

// ============================================================
// 10. DEMO & TEST
// ============================================================

int main() {
    printf("=== DYNAMIC ARRAY / ARRAYLIST DEMO ===\n\n");
    
    // Khởi tạo
    DynamicArray *arr = createArray();
    printArray(arr);
    
    printf("\n--- THÊM PHẦN TỬ ---\n");
    // Thêm vài phần tử (không cần resize)
    push(arr, 10);
    push(arr, 20);
    push(arr, 30);
    printArray(arr);
    
    printf("\n--- TRUY CẬP ---\n");
    printf("Element tại index 0: %d\n", get(arr, 0));
    printf("Element tại index 2: %d\n", get(arr, 2));
    
    printf("\n--- THÊM PHẦN TỬ CHI PHÍ CAO (RESIZE) ---\n");
    // Thêm cho đến khi cần resize
    for (int i = 4; i <= 15; i++) {
        push(arr, i * 10);
    }
    printArray(arr);
    
    printf("\n--- XOÁ CUỐI ---\n");
    pop(arr);
    pop(arr);
    printArray(arr);
    
    printf("\n--- XOÁ TẠI VỊ TRÍ ---\n");
    removeAt(arr, 2);
    printArray(arr);
    
    // Giải phóng
    freeArray(arr);
    
    return 0;
}

/*
 * ============================================================
 * TÓM TẮT ĐỘ PHỨC TẠP
 * ============================================================
 * 
 * | Phép toán   | Độ phức tạp | Ghi chú                    |
 * |-------------|-------------|---------------------------|
 * | Get         | O(1)        | Truy cập trực tiếp         |
 * | Push (cuối) | O(1)*       | * Amortized, resize O(n)   |
 * | Pop (cuối)  | O(1)        |                            |
 * | Insert (k)  | O(n)        | Cần dịch các phần tử sau    |
 * | Remove (k)  | O(n)        | Cần dịch các phần tử sau    |
 * | Search      | O(n)        | Phải kiểm tra từng phần tử  |
 * 
 * ============================================================
 * NETWORK APPLICATION
 * ============================================================
 * 
 * 1. Packet Buffer: lưu các packet đến
 * 2. Socket Table: danh sách các connection
 * 3. Routing Table: danh sách các route (thường dùng hash table bên trên)
 * 
 * ============================================================
 */

File: DataStructutreAndAlgorithm/DataStructure/01_Array_ArrayList.java
/**
 * DATA STRUCTURE: Array / Dynamic Array (ArrayList)
 * 
 * JAVA VERSION
 * 
 * STORY:
 * Trong Java, bạn không cần lo về việc cấp phát bộ nhớ thủ công.
 * ArrayList tự động giúp bạn quản lý việc resize.
 * 
 * So sánh:
 * - C:    Bạn tự quản lý (malloc, free, resize)
 * - Java: Ngôn ngữ quản lý (dù phía sau cũng giống C)
 * 
 * Ưu điểm:
 * - Lập trình nhanh, ít lỗi bộ nhớ
 * 
 * Nhược điểm:
 * - Có Garbage Collection overhead (nhưng thường nhanh hơn việc bạn quản lý)
 */

import java.util.ArrayList;
import java.util.List;

public class DynamicArrayDemo {
    
    /**
     * Ví dụ 1: Sử dụng ArrayList cơ bản
     */
    public static void basicUsage() {
        System.out.println("=== BASIC ARRAYLIST USAGE ===\n");
        
        // Tạo ArrayList
        List<Integer> arr = new ArrayList<>();
        System.out.println("✓ Tạo ArrayList xong\n");
        
        // Thêm phần tử
        System.out.println("--- THÊM PHẦN TỬ ---");
        arr.add(10);
        arr.add(20);
        arr.add(30);
        System.out.println("Array sau khi thêm: " + arr);
        System.out.println("Size: " + arr.size() + "\n");
        
        // Truy cập
        System.out.println("--- TRUY CẬP ---");
        System.out.println("Element tại index 0: " + arr.get(0));
        System.out.println("Element tại index 2: " + arr.get(2) + "\n");
        
        // Xóa
        System.out.println("--- XOÁ ---");
        arr.remove(1);  // Xóa element tại index 1 (giá trị 20)
        System.out.println("Array sau khi xóa index 1: " + arr + "\n");
        
        // Kiểm tra
        System.out.println("--- KIỂM TRA ---");
        System.out.println("Có chứa 30? " + arr.contains(30));
        System.out.println("Vị trí của 30? " + arr.indexOf(30) + "\n");
    }
    
    /**
     * Ví dụ 2: Tạo Custom ArrayList để hiểu rõ hơn cơ chế resize
     * (giống như ở C)
     */
    static class CustomArrayList<T> {
        private Object[] data;
        private int size;
        private static final int INITIAL_CAPACITY = 10;
        
        public CustomArrayList() {
            this.data = new Object[INITIAL_CAPACITY];
            this.size = 0;
        }
        
        /**
         * Thêm phần tử vào cuối
         * 
         * Nếu size == capacity, resize
         */
        public void add(T value) {
            // Kiểm tra nếu cần resize
            if (size == data.length) {
                resize();
            }
            
            // Thêm phần tử
            data[size] = value;
            size++;
        }
        
        /**
         * Resize: Tăng capacity gấp đôi
         * 
         * Mô phỏng:
         * Capacity: 10 → 20 → 40 → 80 → ...
         */
        private void resize() {
            System.out.println("  → Resize! Capacity cũ: " + data.length);
            
            int newCapacity = data.length * 2;
            Object[] newData = new Object[newCapacity];
            
            // Copy dữ liệu cũ
            System.arraycopy(data, 0, newData, 0, size);
            
            data = newData;
            System.out.println("  → Resize xong. Capacity mới: " + data.length);
        }
        
        /**
         * Lấy phần tử tại vị trí index
         */
        @SuppressWarnings("unchecked")
        public T get(int index) {
            if (index < 0 || index >= size) {
                throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
            }
            return (T) data[index];
        }
        
        /**
         * Xóa phần tử tại vị trí index
         */
        @SuppressWarnings("unchecked")
        public T remove(int index) {
            if (index < 0 || index >= size) {
                throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
            }
            
            T value = (T) data[index];
            
            // Dịch các phần tử sau về phía trước
            System.arraycopy(data, index + 1, data, index, size - index - 1);
            
            size--;
            return value;
        }
        
        public int getSize() {
            return size;
        }
        
        public int getCapacity() {
            return data.length;
        }
        
        @Override
        public String toString() {
            StringBuilder sb = new StringBuilder("[");
            for (int i = 0; i < size; i++) {
                sb.append(data[i]);
                if (i < size - 1) sb.append(", ");
            }
            sb.append("]");
            return sb.toString() + " (Size: " + size + ", Capacity: " + data.length + ")";
        }
    }
    
    /**
     * Ví dụ 3: Demo Custom ArrayList
     */
    public static void customArrayListDemo() {
        System.out.println("\n=== CUSTOM ARRAYLIST DEMO ===\n");
        
        CustomArrayList<Integer> arr = new CustomArrayList<>();
        
        System.out.println("--- THÊM PHẦN TỬ ---");
        for (int i = 1; i <= 15; i++) {
            System.out.print("Thêm " + (i * 10) + ": ");
            arr.add(i * 10);
            System.out.println(arr);
        }
        
        System.out.println("\n--- TRUY CẬP ---");
        System.out.println("Element tại index 0: " + arr.get(0));
        System.out.println("Element tại index 5: " + arr.get(5));
        
        System.out.println("\n--- XOÁ ---");
        System.out.println("Xóa index 2 (giá trị " + arr.get(2) + ")");
        arr.remove(2);
        System.out.println("Array sau khi xóa: " + arr);
    }
    
    /**
     * Ví dụ 4: So sánh độ hiệu năng (ArrayList vs Array)
     */
    public static void performanceComparison() {
        System.out.println("\n=== PERFORMANCE COMPARISON ===\n");
        
        int n = 1_000_000;
        
        // Thử nghiệm 1: ArrayList
        System.out.println("--- ArrayList (với resize) ---");
        long start = System.nanoTime();
        List<Integer> arrayList = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            arrayList.add(i);
        }
        long end = System.nanoTime();
        System.out.println("Thời gian thêm " + n + " phần tử: " + (end - start) / 1_000_000.0 + " ms");
        
        // Thử nghiệm 2: Array cố định (phải biết kích thước trước)
        System.out.println("\n--- Array cố định ---");
        start = System.nanoTime();
        int[] array = new int[n];
        for (int i = 0; i < n; i++) {
            array[i] = i;
        }
        end = System.nanoTime();
        System.out.println("Thời gian gán " + n + " phần tử: " + (end - start) / 1_000_000.0 + " ms");
    }
    
    /**
     * Ví dụ 5: Ứng dụng thực tế - Lưu packet buffer
     */
    static class PacketBuffer {
        private List<String> packets = new ArrayList<>();
        
        public void receivePacket(String packet) {
            packets.add(packet);
        }
        
        public String getPacket(int index) {
            return packets.get(index);
        }
        
        public int getPacketCount() {
            return packets.size();
        }
        
        public void printAllPackets() {
            System.out.println("Total packets: " + packets.size());
            for (int i = 0; i < packets.size(); i++) {
                System.out.println("  Packet " + (i + 1) + ": " + packets.get(i));
            }
        }
    }
    
    public static void networkApplicationDemo() {
        System.out.println("\n=== NETWORK APPLICATION DEMO ===\n");
        
        PacketBuffer buffer = new PacketBuffer();
        
        System.out.println("--- Nhận packet ---");
        buffer.receivePacket("HTTP GET request");
        buffer.receivePacket("TCP SYN");
        buffer.receivePacket("UDP DNS query");
        buffer.receivePacket("TCP ACK");
        
        buffer.printAllPackets();
        
        System.out.println("\n--- Truy cập packet cụ thể ---");
        System.out.println("Packet thứ 2: " + buffer.getPacket(1));
    }
    
    /**
     * ============================================================
     * TÓM TẮT ĐỘ PHỨC TẠP
     * ============================================================
     * 
     * | Phép toán   | Độ phức tạp | Ghi chú                    |
     * |-------------|-------------|---------------------------|
     * | Get         | O(1)        | Truy cập trực tiếp         |
     * | Add (cuối)  | O(1)*       | * Amortized, resize O(n)   |
     * | Remove (k)  | O(n)        | Cần dịch các phần tử sau    |
     * | Search      | O(n)        | Phải kiểm tra từng phần tử  |
     * | Insert      | O(n)        | Cần dịch các phần tử sau    |
     * 
     * ============================================================
     */
    
    public static void main(String[] args) {
        basicUsage();
        customArrayListDemo();
        performanceComparison();
        networkApplicationDemo();
    }
}

/*
 * CHI PHÍ RESIZE:
 * 
 * Nếu tăng capacity từng cái một:
 *     1 + 2 + 3 + ... + n = n(n+1)/2 = O(n²)
 * 
 * Nếu tăng gấp đôi:
 *     1 + 2 + 4 + 8 + ... + n = 2n - 1 = O(n)
 *     Amortized: O(n) / n = O(1) per add
 * 
 * Đó là lý do ArrayList tăng gấp 1.5 hoặc 2 lần, không tăng từng cái một.
 */
